### ISSUE TITLE
Authentication support

### ISSUE THREAD

**Issue by edmondchuc** [oxigraph/oxigraph#88]
Hi, fantastic project!

It would be awesome to have authentication (and perhaps authorisation) support for the SPARQL and graph store endpoints in the Oxigraph server. For example, you can have the store as a public facing read-only SPARQL endpoint while the SPARQL update endpoint is protected with basic auth.

I would love to contribute to this project. I don't have any experience in Rust (but this project has inspired me to learn it). I mainly work in Python. 

Some options to implement basic auth:
- NGINX (easiest but probably not user-friendly in terms of deployment)
- Bundle a Python Flask/FastAPI proxy with a Next.js frontend in the Docker image of Oxigraph to handle and manage basic auth 
  - I can help with this, and implement role-based access control and create an admin dashboard similar to RDF4J and GraphDB
- Implement auth in Oxigraph server (in Rust) and use a Next.js frontend for the admin dashboard.

The chosen option will obviously affect future feature implementations such as authorisation. 

If we go with the latter two options, we can use a file-based JSON (or RDF) store. This makes it very easy to deploy and modify things for sys admins (following GraphDB's approach). 

Let me know your thoughts.

**Comment by ansarizafar** [comment]
Authentication and authorization should be built in Oxigraph. I really like how Flureedb is doing authorization with smart functions (currently only clojure is supported but JavaScript support is planned)

**Comment by Tpt** [comment]
Hi! Thank you so much for stepping in @edmondchuc @ansarizafar 

I see two different points here:
1. Authentication/authorization.
2. A nice admin dashboard frontend. It would be much welcome and a nice addition compared to [the current very limited JS interface](https://github.com/oxigraph/oxigraph/blob/master/server/templates/query.html).

About authentication/authorization, I currently use Nginx. This way, I could use all the Nginx plugins to do authentication, filtering, load balancing...
Personally, the feature I miss the most at the moment is the ability to limit the execution time of SPARQL queries to avoid overloading the server with resource-consuming queries.
I fear that adding authentication to Oxigraph means having to add a significant set of authentication providers (static user/password, LDAP...) even if there is currently with Oxigraph only three possible authorization levels: nothing, read and read-write.
I would be very curious about the practical use cases you have for authentication/authorization to see how well we could support them without having to build yet another authentication/authorization system. But if there are use cases where it makes sense to have authentication/authorization inside of Oxigraph server, I am definitely fine to have them integrated there.

An admin dashboard would be very nice and welcome. If I understand your message correctly, there are two approaches here (please, correct me if I am wrong):
1. Static assets served directly by the Rust server just like it's already done for the current simplistic UI. It has the advantage of easy build and deployment but basically requires to use plain JS and CSS and only relies on JS libraries that do not require a build step.
2. Have a separate server for the admin dashboard. But it requires to proxy the Rust server through the dashboard, to have a frontend proxy, or to proxy the dashboard server through the Rust server. It also adds an extra build step (build the dashboard).
Personally, I would be more in favor of option 1 because it would be much simpler to build and deploy. But it adds a lot of constraint on the kind of framework/library that could be use frontend-wise.

I would be more in favor of option 1 because it's easier for deployment, packaging and does not add new dependencies. But if it makes the dashboard significantly more difficult to implement, then option 2 might make more sense.

**Comment by edmondchuc** [comment]
Hi @Tpt, yes looks like this thread originally started on authentication/authorization but quickly got conflated with a nice admin dashboard 😄. 

I understand you favor option 1 but there are some things to think about if we do go this route. Where is auth handled here? In the rust server I assume. In this case, I am not sure how much I can help out since I am not knowledgeable in Rust development at all (currently - I can always learn). 

For me, it will be easier to go with option 2. We can create an MVP much faster with Python (at least for me). If there are performance considerations, we can always re-implement it at a later date back into the Rust server.

The architecture for option 2 will look something like this. ![oxigraph admin architecture](https://i.imgur.com/IpycQLz.png)

Let me know your thoughts.

**Comment by Tpt** [comment]
In my option 1 I see two options for auth.
a. It is not managed at all. The admin dashboard has full access on the database and people who want auth have to restrict the access to the admin dashboard using nginx or something. It's an "admin" dashboard so it might not be a big deal.
b. It would be managed by the Rust server itself.

For an MVP, indeed your option seems good. However, I would like to still allow users to just run Oxigraph without admin UI and proxy from a single Docker container. Running a single 20MB container on top of a Debian image is much simpler than having to deal with a complete network. In this option we would have "oxigraph server" that would be the minimalist Rust server and something like "oxigraph workbench" (just like rdf4j workbench or mysql workbench) that would provide the full set of features. If auth get traction we might take the time to convert it to Rust and move it to the server version.

What do you think about it?

**Comment by ansarizafar** [comment]
Authentication should be part of rust server .

**Comment by Tpt** [comment]
> Authentication should be part of Rust server .

Yes, definitely. However, it seems that @edmondchuc does not feel able to implement it in Rust at the moment and I would prefer to focus myself on features more useful to me first like performance improvements, full test search, [RDF-star](https://w3c.github.io/rdf-star/cg-spec/editors_draft.html) and JSON-LD support. If you want to step in and implement authentication in Rust server I think it would be very much welcomed.

**Comment by alexkreidler** [comment]
Hi all, 

This is a very interesting idea.

Because most of the administration functionality (setting timeouts on queries, some sort of auth) would be managed in Oxigraph in Rust, I don't see much of a reason to write a separate admin backend in Python, as it would probably just be forwarding the requests. There could be good reasons for separating them though, I just can't think of them right now.

My preferences for the frontend would be something like a React Next.js app that gets exported to static HTML + JS. This would then connect to whatever endpoint is configured in the UI's settings (could be `localhost`, `0.0.0.0`, or a remote endpoint)

This could be served either by:
- bundling the static files with the Rust binary (the downside is that the DB version and app version would be tied, which is probably not a great thing)
- any web server the user wants (e.g. nginx in a Docker container)
- on a static web hosting service (free ones like Netlify or Github pages would work) as long as Oxigraph adds CORS headers that allow those connections. The static web hosting would mean that we could automatically push updates from CI/CD, and the user wouldn't have to maintain anything.

In a few weeks I'll have some time to maybe write an admin UI if we can decide on what the functionality should be. I'd also agree with @Tpt that we should focus on the minimal set of features to start. I think query timeouts is probably good to start with.

**Comment by Tpt** [comment]
> Because most of the administration functionality (setting timeouts on queries, some sort of auth) would be managed in Oxigraph in Rust, I don't see much of a reason to write a separate admin backend in Python.

I agree. However I personally don't have these features very high on my personal TODO list, so if someone wants to implement them in a proxy because they fill unable to do them in Rust, it is fine for me as a temporary solution.

+1 To be able to build a "static" app.  Having it would not prevent any of the 3 usecases you suggest.

> In a few weeks I'll have some time to maybe write an admin UI if we can decide on what the functionality should be. 

Thank you! It would be amazing! For me, the most important feature is to be able to easily send query and updates from the UI and get some idea of which data are in there. But it might require to solve #94 first to be efficient.

**Comment by MarcAntoine-Arnaud** [comment]
Hello,

I'm starting using oxigraph, and as I'm also involved on other projects that use Solid, I think it can be great to look on:
https://solidproject.org/TR/oidc

(and maybe useful in the future to provide a compliant API to Solid project)

**Comment by jonassmedegaard** [comment]
Regarding Solid-OIDC, there is [existing work](https://lib.rs/crates/manas_authentication).

**Comment by JeancarloBarrios** [comment]
Just pinging to see if this was eve developed. If not I don't mind contributing to it. Should I start a small ADR? or something?
